### PR TITLE
Add File Naming setting for more detailed naming

### PR DESCRIPTION
**Pull request by m1-ins**
# What's new

Solved Issue #2598 (Feature Request) by adding more detail to file names through the use of a non-invasive setting named 'File Naming'.

# Verification 

1. Go to: Settings > System > File Naming > Set to 'Detailed'
2. Create any file that is an internal app which supports the naming convention (Sub-Ghz, NFC, iButton, RFID)
3. Save the file

You can check the difference by setting the filename mode to 'Default'.

# Checklist (For Reviewer)

- [x] PR has description of feature/bug or link to Confluence/Jira task
- [x] Description contains actions to verify feature/bugfix
- [x] I've built this code, uploaded it to the device and verified feature/bugfix

### CODE REVIEW

File: firmware/targets/furi_hal_include/furi_hal_rtc.h [hunk 1]
```diff
@@ -31,7 +31,8 @@ typedef enum {
     FuriHalRtcFlagC2Update = (1 << 3),
     FuriHalRtcFlagHandOrient = (1 << 4),
     FuriHalRtcFlagLegacySleep = (1 << 5),
-    FuriHalRtcFlagStealthMode = (1 << 6),
+    FuriHalRtcFlagFilenameMode = (1 << 6),
```
Comment (skotopes) [inline firmware/targets/furi_hal_include/furi_hal_rtc.h]: Nope, you can not replace already existing flag

File: lib/toolbox/set_name.c [hunk 2]
```diff
@@ -0,0 +1,19 @@
+#include "set_name.h"
+#include <stdio.h>
+#include <stdint.h>
+#include <furi_hal_rtc.h>
+
+void set_name(char* name, uint8_t max_name_size, char* app_name) {
```
Comment (skotopes) [inline lib/toolbox/set_name.c]: I'd rather see one universal interface for random and detailed api with rtc flag check on the inside

File: firmware/targets/furi_hal_include/furi_hal_rtc.h (hunk 1 above)
Comment (skotopes) [reply firmware/targets/furi_hal_include/furi_hal_rtc.h]: Also mode in FuriHalRtcFlagFilenameMode kinda expects that there will be some options. Not sure how it should be called.

File: firmware/targets/f7/api_symbols.csv [hunk 3]
```diff
@@ -1,5 +1,5 @@
 entry,status,name,type,params
-Version,+,35.0,,
+Version,+,36.0,,
```
Comment (skotopes) [inline firmware/targets/f7/api_symbols.csv:2]: Major bump is not needed here

File: applications/main/ibutton/scenes/ibutton_scene_save_name.c [hunk 4]
```diff
@@ -17,7 +19,11 @@ void ibutton_scene_save_name_on_enter(void* context) {
     const bool is_new_file = furi_string_empty(ibutton->file_path);
 
     if(is_new_file) {
-        set_random_name(ibutton->key_name, IBUTTON_KEY_NAME_SIZE);
+        if(furi_hal_rtc_is_flag_set(FuriHalRtcFlagFilenameMode)) {
```
Comment (skotopes) [inline applications/main/ibutton/scenes/ibutton_scene_save_name.c]: It would be nice to see universal interface to generate names here

**Review by skotopes — CHANGES_REQUESTED** [review]

File: firmware/targets/furi_hal_include/furi_hal_rtc.h (hunk 1 above)
Comment (m1-ins) [reply firmware/targets/furi_hal_include/furi_hal_rtc.h]: Leaving this open, as I am unsure if this change is visual / the right fix. I made sure to keep FuriHalRtcFlagStealthMode, but added FuriHalRtcFlagFilenameScheme in the most recent commit.

File: lib/toolbox/name_generator.c [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+const char* convert_app_extension_to_name(char* app_name) {
+    struct NameConversion {
+        const char* original;
+        const char* converted;
+    };
+
+    struct NameConversion conversions[] = {
+        {".ibtn", "iBtn"}, {".nfc", "NFC"}, {".rfid", "RFID"}, {".sub", "RAW"}};
```
Comment (skotopes) [inline lib/toolbox/name_generator.c]: That doesn't looks extensible. Why not to pass it explicitly from application side?

File: applications/main/subghz/scenes/subghz_scene_save_name.c [hunk 6]
```diff
@@ -40,7 +41,9 @@ void subghz_scene_save_name_on_enter(void* context) {
 
     if(!subghz_path_is_file(subghz->file_path)) {
         char file_name_buf[SUBGHZ_MAX_LEN_NAME] = {0};
-        set_random_name(file_name_buf, SUBGHZ_MAX_LEN_NAME);
+
+        name_generator_set(file_name_buf, SUBGHZ_MAX_LEN_NAME, SUBGHZ_APP_EXTENSION);
```
Comment (skotopes) [inline applications/main/subghz/scenes/subghz_scene_save_name.c]: That is not raw

**Review by skotopes — CHANGES_REQUESTED** [review]

**Comment by skotopes** [comment]
Un-draft when ready

**Comment by m1-ins** [comment]
I believe I have addressed the issues raised in the last code review, but if there is logic/ the implementation that needs be changed specifically within the `convert_app_extension_to_name` function, let me know!

File: lib/toolbox/name_generator.h [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+ * @param max_name_size length of given buffer
+ * @param app_name name of the application
+ */
+void name_generator_set(char* name, uint8_t max_name_size, const char* app_name);
+
+/** Converts application extension to app_name
+ * @param app_name name of the application
+ * @return converted name
+ */
+const char* convert_app_extension_to_name(char* app_name);
```
Comment (skotopes) [inline lib/toolbox/name_generator.h]: There is no need in this method. Application that uses name generator already knows its name, just provide there.

File: lib/toolbox/name_generator.c [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+    struct NameConversion {
+        const char* original;
+        const char* converted;
+    };
+
+    struct NameConversion conversions[] = {
+        {".ibtn", "iBtn"}, {".nfc", "NFC"}, {".rfid", "RFID"}, {".sub", "SubGhz"}};
+
+    const int numConversions =
+        sizeof(conversions) / sizeof(conversions[0]); // gets the number of potential conversions
```
Comment (skotopes) [inline lib/toolbox/name_generator.c]: COUNT_OF()

**Review by skotopes — APPROVED** [review]

**Pull request merged**